### PR TITLE
feat(parser): `DynamoDBMarshalled` helper to parse DynamoDB data structure

### PR DESCRIPTION
**Pull request by arnabrahman**
## Summary
When working with Amazon DynamoDB Stream events, the service sends the DynamoDB items in the database-specific format rather than regular JSON objects.

For example, a string might be represented as { "S": "foo" } rather than just foo. This makes parsing the actual changes within an event hard to parse and validate since it requires customers writing a Zod schema with this in mind.

This PR introduces a helper function `DynamoDBMarshalled` to work this structure
### Changes

- Helper function to unmarshall and validate dynamoDB stream schema
- New helper function is exported to a subpath
- Unit test the new helper function
- Write up the docs for the new helper function

**Issue number:** #3194 

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

File: packages/parser/src/helpers/dynamodb.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+ *     }
+ *   ]
+ * }
+ * ```
+ *
+ * @param schema - The schema to validate the JSON string against
+ */
+const DynamoDBMarshalled = <T extends ZodTypeAny>(schema: T) =>
+  z
+    .record(z.string(), z.custom<AttributeValue>())
```
Comment (arnabrahman) [inline packages/parser/src/helpers/dynamodb.ts:66]: I would appreciate suggestions on using this type.

Setting the type to `unknown` causes a TypeScript error. The `@aws-sdk/util-dynamodb` package has a peer dependency on `@aws-sdk/client-dynamodb` and we are using the `AttributeValue` type from `@aws-sdk/client-dynamodb`.  Should we also declare `@aws-sdk/client-dynamodb` as a peer dependency? 

Alternatively, we could set the type to `unknown` and suppress the TypeScript error using `biome-ignore`, though I would prefer to avoid this approach.

Comment (dreamorosi) [reply packages/parser/src/helpers/dynamodb.ts:66]: I was on the fence and I shared the concern, but after thinking about it I think using the type from `@aws-sdk/client-dynamodb` is fine, anyway you'd need these dependencies only if you're actually working with this helper.

I would however change the helper from this:

```ts
z.record(z.string(), z.custom<AttributeValue>())
```

to this:

```ts
z.union([
  z.custom<Attributevalue>(),
  z.record(z.string(), z.custom<AttributeValue>())
])
```

Since the `unmarshall()` function has this signature `Record<string, AttributeValue> | AttributeValue`.

**Review by dreamorosi — CHANGES_REQUESTED** [review]
Thank you for the PR and for your patience!

I have a minor comment in the signature of the helper so that it's aligned with the underlying `unmarshall` function, but other than that I am ready to approve the PR.

File: packages/parser/src/helpers/dynamodb.ts (hunk 1 above)
Comment (arnabrahman) [reply packages/parser/src/helpers/dynamodb.ts]: One thing, should we also add `@aws-sdk/client-dynamodb` as a peer dependency? 

The reason is that if someday, `@aws-sdk/util-dynamodb` decides to use its own type for `unmarshall` rather than `AttributeValue` as this is the only use case for `client-dynamodb` inside `@aws-sdk/util-dynamodb` and removes `@aws-sdk/client-dynamodb` from its peer dependencies, we might encounter errors here.

I don't know I might be overthinking here.

**Review by dreamorosi — APPROVED** [review]
Thank you for the PR!

**Comment by dreamorosi** [comment]
Just as FYI @arnabrahman - Over the next couple of weeks the team and I will invest time in creating a roadmap and backlog for the next two quarters.

With that, we'll be creating a number of issues and label them accordingly.

**Comment by arnabrahman** [comment]
@dreamorosi Thanks for the update, will be looking forward to it 🚀
